### PR TITLE
SONARJS-805 Rule: Non-existent properties shouldn't be accessed for reading

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java [hunk 1]
```diff
@@ -72,7 +76,11 @@ public Type parentType() {
 
   private SymbolicValue getValueFromPrototype(String propertyName) {
     if (parentType != null) {
-      return parentType.getValueForProperty(propertyName);
+      SymbolicValue valueForProperty = parentType.getValueForProperty(propertyName);
+      if (valueForProperty == UnknownSymbolicValue.UNKNOWN && PRIMITIVE_TYPES.contains(this)) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java [hunk 2]
```diff
@@ -51,6 +54,7 @@
   ;
 
   private static final List<Type> VALUES_REVERSED = Lists.reverse(Arrays.asList(Type.values()));
+  private static final Set<Type> PRIMITIVE_TYPES = ImmutableSet.of(NUMBER, STRING, BOOLEAN);
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java]: MAJOR Convert this Set to an EnumSet. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1641)

Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/se/Type.java]: MAJOR Convert this Set to an EnumSet. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1641)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonExistentPropertyAccessCheck.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      IdentifierTree propertyTree = ((DotMemberExpressionTree) element).property();
+
+      if (!treesWithIssues.contains(propertyTree)) {
+        addIssue(element, String.format(MESSAGE, propertyTree.name(), getTypeOfObject(programStateBefore)));
+        treesWithIssues.add(propertyTree);
+      }
+    }
+  }
+
+  private String getTypeOfObject(ProgramState programState) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonExistentPropertyAccessCheck.java]: MAJOR Make "getTypeOfObject" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/NonExistentPropertyAccessCheck.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  private static String getTypeOfObject(ProgramState programState) {
+    Constraint constraintOnObject = programState.getConstraint(programState.peekStack());
+    Type type = Type.find(constraintOnObject);
+    Preconditions.checkState(type != null && TYPE_NAMES.containsKey(type));
+
+    return TYPE_NAMES.get(type);
+  }
+
+  private static boolean isUndefinedProperty(ProgramState currentState, Tree element) {
+    return element.is(Kind.DOT_MEMBER_EXPRESSION) && SpecialSymbolicValue.UNDEFINED.equals(currentState.peekStack());
```
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/NonExistentPropertyAccessCheck.java]: @vilchik-elena Suggested name: `isUndefinedPropertyAccess`

**Pull request merged**